### ISSUE TITLE
`LRScheduler` Refactoring

### ISSUE THREAD

**Issue by GdoongMathew** [pytorch/pytorch#125531]
## Feature
Proposal to solve issue mainly raised from issue #67760, and some other issues related to `lr_scheduler` module.

* Refactor `lr_scheduler`  module
  * make `lr_scheduler` class to support all hyperparameter settings in the given `optimizer`
  * refactor `SequentialLR` & `ChainedScheduler` to support `ReduceLROnPlateau` with arbitrary step kwargs inputs.
    * #110761 
    * #68978
* Deprecate `step(epoch)` warnings and remove supports of `epoch`, and *probably change to closed-form implementation* as well.
  * conflicts: 
    * `Secheduelr` composed by a `SequentialLR`, will raise #119168. Closed-form implementation will cause every .
    * Currently, `CyclicLR` is missing closed-form implementation, as well as `ReduceLROnPlateau`.
    * Unittests for `SequentialLR` expect all schedulers to use close-form implementation, whereas the `ChainedScheduler` is not.
  * #55135
  * #67586
  * #116776
  * #67958
* Allow `param_groups` to be passed to `lr_scheduler` class
  * #72146
  * probable issues:
    * `param_groups` must be an instance of `dict` right from the beginning, or else any additional keys added by the `optimizer` will not be accessible by the `lr_scheduler` class.
    * Currently, there's no way to validate whether the `param_groups` passed to the `lr_scheduler` class are all in the same `optimizer` instance or not.
* Deprecate `print_lr`
  * #99270

## Proposed Changes
1. A new abstract class `_SchedulerBase` will be created.
```python
class _SchedulerBase(abc.ABC):
    optimizer: Optimizer

    @abc.abstractmethod
    def step(self, **kwargs):
        raise NotImplementedError

    @abc.abstractmethod
    def state_dict(self): ...

    @abc.abstractmethod
    def load_state_dict(self): ...
```
2. Reneme `lr_scheduler` to `Scheduler` and refactor it to support all hyperparameter settings in the given `optimizer`.
```python
class Scheduler(_SchedulerBase):
    @property
    def targets(self) -> Sequence[str]:
        """The targets that the schedulers can update."""
        ...

    @property
    def last_targets(self) -> Sequence[Dict[str, Any]]:
        """The targets at the last step for each parameter group."""
        ...

    def update_targets(self, **kwargs) -> None: 
        """Update the targets (lr, momentum, ...) for each parameter group."""
        ...
    
    def step(self, epoch: Optional[int] = None, **kwargs) -> None:
        """Update the learning rate of each parameter group."""
        if epoch is not None:
            warnings.warn("The epoch parameter is deprecated and will be removed in the future.", DeprecationWarning)
            self.last_step = epoch
        else:
          self.last_step += 1
        
        with _enable_get_lr_call(self):
            self.update_targets(step=self.last_step, **kwargs)
```
3. Refactor `SequentialLR` & `ChainedScheduler` to support `ReduceLROnPlateau` with arbitrary step kwargs inputs.
```python
class ComposedScheduler(_SchedulerBase, abc.ABC):
    """ChainedScheduler and SequentialLR can subclass this."""

    def __init__(
            self,
            schedulers: Sequence[LRScheduler],
            last_step: int = -1,
    ): ...

    def step(self, epoch: Optional[int] = None, **kwargs):
      # check epoch deprecation
      self.step_schedulers(step=self.last_step, **kwargs)

    @abc.abstractmethod
    def step_schedulers(self, **kwargs): ...

    @property
    def optimizer(self) -> Optimizer:
      return self.schedulers[0].optimizer

```

I like to spend some time fixing some of the issues regarding to the current scheduler module, and feel free to discuss with me further.

### Alternatives

leave it as it is.

### Additional context

Current development status: [link](https://github.com/GdoongMathew/pytorch/blob/refactor/lr_sch_interface/torch/optim/lr_scheduler.py)

**Comment by vadimkantorov** [comment]
Related discussion:
- https://github.com/pytorch/pytorch/issues/68332

**Comment by GdoongMathew** [comment]
Hi @vadimkantorov, thanks for the comment.
I do think that a stateless or closed-form implementation should be more elegant, and easier to understand for other developers. However, there're a few reasons that I cannot 100% agree on fully moving to close-form schedulers.
1. I tend to think that lr (or any hyperparameters) changing should be smooth. And this becomes an issue if we're talking about `SequentialLR` and `ChainedScheduler` cause both of this will force the next scheduler to update the lr based on the initial value instead of the value from the last iteration, which can be avoided if using non close-form.
2. As mentioned above, `CyclicLR` is still missing a close-form implementation.

**Comment by vadimkantorov** [comment]
I'm not proposing deprecating/not developing stateful schedulers. I'm suggesting that they should be complemented by closed-form, simple functions (and these functions would require much less API design, so can be merged relatively easy).

And you're right - the more complex concepts like chaining maybe cannot be fully expressed with them, but it's okay. It's already beneficial if the "building blocks" are available in closed-form, simple functions fully determined by their function arguments.

**Comment by GdoongMathew** [comment]
Oh, sorry for misinterpreting what you said. Probably I've misunderstood the meaning of `close-form` as well. By reading some of the old issues (I could not really find it now) and the current method naming, I've always thought that in scheduler, `close-form` means any implements which only takes `step` and `initial_lr` as the arguments. 

But after reading the issue you mentioned (#68332), I think one of the way to adapt that into the current proposal, would be something like the following?
```python
def step_lr(lr: float, gamma: float) -> float:
    return lr * gamma

class StepLR(Scheduler):
    def update_targets(self, step: int, **kwargs):
        for param_group in self.param_groups:
            param_group["lr"] = step_lr(param_group["lr"], self.gamma)
```

**Comment by vadimkantorov** [comment]
I originally meant closed-form actually accepting the iteration counts (and many schedules are more intuitively formulated/parameterized using iteration counts for cut-offs - and this is also a parametrization typically used in optimization literature notation, right?), from which the lr can be recovered or reset.

Then the user is already free to do experimentation and custom stuff with resetting the iteration counter to some other value when finetuning from checkpoint

**Comment by janeyx99** [comment]
Thank you for writing up this thorough overview of the LRScheduler landscape, and I do agree there is a big refactor that would make it better. Unfortunately we don't have an active maintainer on LRScheduler today so it may take a few months before we have the time and expertise to review the changes necessary here such that they can be landed in a safe way.
